### PR TITLE
grpc: propagate tracing headers when using google grpc streams

### PR DESCRIPTION
**Pull request by cainelli**
Commit Message: grpc: propagate tracing headers when using google grpc streams
Additional Description:  Following up on https://github.com/envoyproxy/envoy/pull/33665, which adds tracing to Envoy gRPC streams, this PR introduces the same support using Google gRPC for completeness to https://github.com/envoyproxy/envoy/issues/21119.

Risk Level: Low
Testing: The same manual tests described in https://github.com/envoyproxy/envoy/pull/33665, besides the integration tests of extproc are also testing google grpc now.
Docs Changes: N/A
Release Notes: N/A

### CODE REVIEW

File: source/common/grpc/google_async_client_impl.cc
Comment (htuch) [inline source/common/grpc/google_async_client_impl.cc:171]: Nit: prefer the other way around in Envoy style. I know why this is done historically for C-derived languages, but it's quite jarring when the rest of the code base doesn't follow this style and modern compilers are fairly good at picking up mistakes here.
```suggestion
  if (options.parent_span_ != nullptr) {
```

File: source/common/grpc/google_async_client_impl.cc
Comment (htuch) [inline source/common/grpc/google_async_client_impl.cc:174]: Where does this naming scheme come from, e.g. why is this "egress"?

**Review by htuch — COMMENTED** [review]
Looks reasonable, @tyxia can you PTAL as well? Thanks.

File: source/common/grpc/google_async_client_impl.cc
Comment (cainelli) [reply source/common/grpc/google_async_client_impl.cc]: TIL about Yoda conditions, thank you. I'm new to the language :)

File: source/common/grpc/google_async_client_impl.cc
Comment (cainelli) [reply source/common/grpc/google_async_client_impl.cc:174]: I used the same as unary grpc calls were using [[1]](https://github.com/envoyproxy/envoy/blob/99bb5659f8b204068cdb307a898d617d0ee4b5e8/source/common/grpc/google_async_client_impl.cc#L447-L449)[[2]](https://github.com/envoyproxy/envoy/blob/99bb5659f8b204068cdb307a898d617d0ee4b5e8/source/common/grpc/async_client_impl.cc#L301-L303) and also what was added in envoy grpc stream [[3]](https://github.com/envoyproxy/envoy/blob/99bb5659f8b204068cdb307a898d617d0ee4b5e8/source/common/grpc/async_client_impl.cc#L92-L95), let me know if there are guidelines to follow here.

File: source/common/grpc/google_async_client_impl.cc
Comment (tyxia) [inline source/common/grpc/google_async_client_impl.cc:178]: [GoogleAsyncRequestImpl](https://github.com/envoyproxy/envoy/blob/99bb5659f8b204068cdb307a898d617d0ee4b5e8/source/common/grpc/google_async_client_impl.cc#L450) is also setting  UpstreamCluster  ----- `current_span_->setTag(Tracing::Tags::get().UpstreamCluster, parent.stat_prefix_)`

I am not able to find out why `UpstreamCluster` was originally set for google gRPC as it doesn't make use of  Envoy’s  upstream connection management like envoy-gRPC.   + @wbpcode / @htuch in case they know its use case.

We probably should have consistent behavior here: EITHER add UpstreamCluster if it does have value OR remove it in googleAsyncRequest

**Review by tyxia — COMMENTED** [review]
LGTM modulo the comment below, Thanks!

@cainelli  Please don't do force-push in the future as we want to keep the full commit history

File: source/common/grpc/google_async_client_impl.cc
Comment (htuch) [reply source/common/grpc/google_async_client_impl.cc:178]: Can we factor this to a common library for span annotations so we keep the two gRPC implementations in sync?

Comment (tyxia) [reply source/common/grpc/google_async_client_impl.cc:179]: @cainelli if you could add a TODO for refactor and follow this up, I am personally fine with this PR to unblock you unless @htuch has strong opinion on this.
 
For common library, we can start with something simple like utility functions/class etc.  The key is singe source-of-truth  (when applicable) for both Envoy-gRPC and google-gRPC

**Comment by tyxia** [comment]
/wait-any

@cainelli  Please also look at the CI failure. Looks like they are legit.

File: source/common/grpc/google_async_client_impl.cc
Comment (cainelli) [reply source/common/grpc/google_async_client_impl.cc:180]: Thanks @tyxia, I added a comment and fixed the test failing.

**Review by tyxia — APPROVED** [review]
LGTM, Thanks!

Defer to @htuch, this open thread in particular https://github.com/envoyproxy/envoy/pull/34395/files#r1624502572

File: source/common/grpc/google_async_client_impl.cc
Comment (htuch) [reply source/common/grpc/google_async_client_impl.cc:180]: OK, this is fine, it's not critical, so a TODO works.

**Review by htuch — APPROVED** [review]
LGTM, thanks!

**Pull request merged**
